src/mpi/coll/include/coll_types.h

```diff
@@ -0,0 +1,90 @@
+/* -*- Mode: C; c-basic-offset:4 ; indent-tabs-mode:nil ; -*- */
+/*
+ *  (C) 2006 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ *  Portions of this code were written by Intel Corporation.
+ *  Copyright (C) 2011-2017 Intel Corporation.  Intel provides this material
+ *  to Argonne National Laboratory subject to Software Grant and Corporate
+ *  Contributor License Agreement dated February 8, 2012.
+ */
+#ifndef MPIC_TYPES_H_INCLUDED
+#define MPIC_TYPES_H_INCLUDED
+
+#include "queue.h"
+
+#define MPIC_FLAG_REDUCE_L 1
+#define MPIC_FLAG_REDUCE_R 0
+
+struct MPIR_Comm;
+struct MPIR_Op;
+
+#ifndef container_of
+#define container_of(P,T,F) \
+    ((T*) ((char *)P - offsetof(T,F)))
+#endif
+
+#define MPIC_Assert MPIR_Assert
+
+#define MPIC_COMM_DECL MPIC_comm_t coll;
+#define MPIC_REQ_DECL MPIC_req_t coll;
+
+#define MPIC_COMM(comm) (&((comm)->coll))
+#define MPIC_REQ(req) (&((req)->coll))
+
+#ifdef MPIC_DEBUG
+#define MPIC_DBG(...) do {MPL_DBG_MSG_FMT(MPIR_DBG_COLL,VERBOSE,(MPL_DBG_FDEST,__VA_ARGS__));} while (0)
+#define MPIC_INLINE static
+#else
+#define MPIC_DBG(...)
+#define MPIC_INLINE MPL_STATIC_INLINE_PREFIX
+#endif
```

Comment:
Is this left-over debugging code?  Defining `MPIC_INLINE` to `static` seems weird.  If it's for debugging, then I'd recommend setting `MPL_STATIC_INLINE_PREFIX` to `static` instead of `static inline` so it's applicable to all uses of the inlining macro, and not just the collectives code.  The `MPIC_INLINE` macro should just go away, AFAICT.

Reply:
Also, is there a reason `MPL_STATIC_INLINE_SUFFIX` is not used as well?

Reply:
I am looking into this. There should be a common solution to this with the rest of the code.

A grep on MPL_STATIC_INLINE_SUFFIX shows that this is being used only in function definitions but not in function implementations where ever there is MPL_STATIC_INLINE_PREFIX. Can you guide us if we should use it everywhere?

Reply:
The main thing with MPIC_INLINE is that it does not have ATTRIBUTE((always_inline)) when debugging. We may want to do the same thing with MPL_STATIC_INLINE_PREFIX.